xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,721 @@
+      }
+    }
+
+    private void close(Exception error) {
+      if (closed) {
+        return;
+      }
+      checkState(pendingAckRequest == null, "Severe bug: pending ACK/NACK not sent");
+      closed = true;
+      cleanUp();
+      requestWriter.onError(error);
+    }
+
+    private void cleanUp() {
+      if (adsStream == this) {
+        adsStream = null;
+      }
+    }
+
+    private void sendLdsRequest(String resourceName) {
+      checkState(requestWriter != null, "ADS stream has not been started");
+      DiscoveryRequest request =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(ldsVersion)
+              .setNode(node)
+              .addResourceNames(resourceName)
+              .setTypeUrl(ADS_TYPE_URL_LDS)
+              .setResponseNonce(ldsRespNonce)
+              .build();
+      requestWriter.onNext(request);
+    }
+
+    private void sendRdsRequest(String resourceName) {
+      checkState(requestWriter != null, "ADS stream has not been started");
+      rdsResourceName = resourceName;
+      DiscoveryRequest request =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(rdsVersion)
+              .setNode(node)
+              .addResourceNames(resourceName)
+              .setTypeUrl(ADS_TYPE_URL_RDS)
+              .setResponseNonce(rdsRespNonce)
+              .build();
+      requestWriter.onNext(request);
+    }
+
+    /**
+     * Prepares a DiscoveryRequest containing the given information as an ACK.
+     */
+    private void prepareAckRequest(String typeUrl, String resourceName, String versionInfo,
+        String nonce) {
+      pendingAckRequest =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(versionInfo)
+              .setNode(node)
+              .addResourceNames(resourceName)
+              .setTypeUrl(typeUrl)
+              .setResponseNonce(nonce)
+              .build();
+    }
+
+    /**
+     * Sends the previously prepared ACK DiscoveryRequest to management server. Updates the latest
+     * accepted version for the corresponding resource type.
+     */
+    private void sendPendingAckRequest() {
+      checkState(pendingAckRequest != null, "There is no pending ACK request");
+      checkState(requestWriter != null, "ADS stream has not been started");
+      // Update the latest accepted version.
+      String typeUrl = pendingAckRequest.getTypeUrl();
+      if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+        ldsVersion = pendingAckRequest.getVersionInfo();
+      } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+        rdsVersion = pendingAckRequest.getVersionInfo();
+      }
+      // TODO: cases for (VHDS/)CDS/EDS.
+      requestWriter.onNext(pendingAckRequest);
+      pendingAckRequest = null;
+    }
+
+    /**
+     * Cancels the pending ACK request and sends its corresponding (with version reverted)
+     * NACK request.
+     */
+    private void nackPendingAckRequest() {
```

Comment:
nit: how about calling it `nackLastResponse` ?

Reply:
The spec for this method is: "cancel the pending ACK, send an NACK instead".  `nackLastResponse` doesn't fully describe this, although I do agree `nackPendingAckRequest` sounds a bit awkward.

Reply:
Outdated as refactored and combined with other methods.